[dazaga_gourane_karra] initial submission

Rieselhilfe opened the pull request.

keyman-server commented:
Thank you for your pull request. You'll see a "build failed" message until the Keyman team has reviewed the pull request and manually initiated the build process.

Every change committed to this branch will become part of this pull request. When you have finished submitting files and are ready for the Keyman team to review this pull request, please post a "Ready for review" comment.

DavidLRowe commented:
Thanks for your submission. This keyboard is in very good shape (so don't be alarmed at all my comments!).

The way that Keyman stores a keyboard project has changed since Keyman Developer version 16 (which was used to create this keyboard) and the format of the keyboards in the repository has been upgraded to match. If you upgrade to [Keyman Developer version 18](https://keyman.com/developer/features) and open your project, you'll be invited to upgrade to the new format.
![image](https://github.com/user-attachments/assets/5c94f1e0-bc98-4810-9b8e-7c36fe2d5187)
This format removes the list of files from the .kpj file and removes the .keyboard_info file. The information that was in the .keyboard_info file is now handled in the Packaging section (in the Keyman package source, that is, .kps file).

Currently in the .kps file you have `© 2023 Zacharias van Stek` whereas the LICENSE.md file has `© 2025 Zacharias van Stek & Henri Schellberg`. We encourage you to drop the date from the copyright statement in the .kps file so that it doesn't have to be updated when the keyboard is revised. The name(s) listed in both places should match. Perhaps you want `© 2023-2025 Zacharias van Stek & Henri Schellberg` in the LICENSE.md file, then `© Zacharias van Stek & Henri Schellberg` in the .kps file (as you have in the .kmn file).

The language code should be "dzg". This is equivalent to "dzg-Latn-TD", but the script and region for "dzg" default to "Latn" and "TD", so the industry standard is to use only "dzg".

Some more minor changes:

There will be other details in the Packaging section to fill in for the new format.

In the readme.htm file, please change the description from `dazaga_gourane 1.0 generated from template.` to something more meaningful to the user about to install the keyboard, perhaps `This keyboard layout is designed for the Dazaga Gourane Karra language.` (And you could include languages other than English if you wanted.)

For the "&TARGETS" in the .kmn file (a list of tick boxes in Keyman Developer), we prefer that you use `any` only (and leave all the others unselected). This is equivalent to selecting all the boxes, but is easier to see the intent.

Currently the touch layout includes both a phone layout and a tablet layout, which seem to be identical. If they are indeed identical, we recommend deleting one of them. Keyman will use a phone layout on a tablet if no tablet layout is available and a tablet layout on a phone if no phone layout is available. Making changes to both layouts and keeping them aligned is more than double the work. 

Observations:

The `&MnemonicLayout` feature is supported on Windows and paritally on Mac (and maybe Linux). At some point in the future, you may want to consider making multiple keyboards, one based on the French AZERTY layout, another on the German QWERTZ layout, another on the English QWERTY layout. You could then drop the MnemonicLayout (or keep it just for one of them) and have a touch layout that corresponds to the desktop layout.

Do you need to type á on the touch layout? I can do it by using the apostrophe (longpress key on ".") followed by "a" and relying on the desktop deadkey rules, but that seems rather cumbersome. Would you want to add accented versions of "a" as longpress keys on the "a" key?

Rieselhilfe commented:
Thanks for the help! I have updated the keyboard and hope I understood it all correctly. Regarding the comments: We've also discussed creating non-mnemonic keyboards for different layouts, and probably we'll do that in the future. And the á is not needed for the dazaga gourane karra script, I've just added it in the desktop layout for symmetry.

DavidLRowe commented:
Thanks for this update. No need to add a-acute if it's not needed.

The dazaga_gourane_karra.php file provides the online help. The welcome.htm provides the help file packaged with the keyboard (available offline). Unfortunately they use slightly different formats, which means some duplication (until such point that we allow both to be generated from the same source file). Specifically we need a second copy of phone.jpg in the source folder and a corresponding change in the welcome.htm file to reference the file as `src="phone.jpg"` (rather than `src="help/phone.jpg"`), since all the files will be at the same level in the package.

The LICENSE.md file needs the word "Copyright" at the beginning of the copyright statement, so that line should read `Copyright © 2023-2025 Zacharias van Stek & Henri Schellberg`. (This is to satisfy our automated build system, which checks for this.)

The dazaga_gourane_karra.keyboard_info should be deleted (removed from this pull request).

See my previous comments about &TARGETS and about removing one of the two nearly identical touch layouts. Currently the "tablet" and "touch" layouts are nearly the same, with some differences in the width and spacing of the keys. It's painful to keep the two layouts synchronized and can cause confusion when debugging. (If at some point in the future, you decide to add a substantially different layout, it's very easy to add a new one, even copying from an existing one.)

Let me know if anything is unclear.

Rieselhilfe commented:
Sorry, it seems my last changes hadn't had saved, thats why the tablet layout was still present and the target list was still convoluted. It should all be worked in now. Thank you so much!

DavidLRowe commented:
Your changes to the keyboard files look good. There are some comments from the compiler about rules that use "´" as a key. Since this character U+00B4 is not produced by the keyboard (at least the QWERTY layout), this rule won't ever be activated. Eventually you may want to see if that rule needs to be modified or just removed, but that doesn't need to be addressed at this point if you are happy with the functionality.

Now there are some things to fix in the packaging.

Currently your list of files includes an incorrect name and fails to include some needed files. With dazaga_gourane_karra.kps selected, "Files" tab on left, then "Remove" or "Add" button at the bottom of the list of files.
- remove dazaga_gourane.js (incorrect name)
- from the build folder, add dazaga_gourane_karra.js (this provides the touch layout)
- from the build folder, add dazaga_gourane_karra.kmx (this provides the desktop layout)
- from the source folder, add phone.jpg (needed by welcome.htm)
- from the keyboard project folder (one level up from the source folder), add LICENSE.md

Normally, we also ask for an on-screen keyboard (OSK), but that is less helpful for mnemonic layouts.

Thanks!

DavidLRowe commented:
One other thing in Packaging: With dazaga_gourane_karra.kps selected, "Keyboards" tab on left, then "Add" button below the box with heading "BCP 47 tag". Enter 'dzg' for the 'Language tag' field (assuming I have the correct code!) and leave the 'Script' and 'Region' fields empty (to use the default Latn and Chad values). You can edit the default 'Language name' field. Then OK.

Rieselhilfe commented:
I worked in your comments. Thanks!

DavidLRowe commented:
Note that I needed to make some changes to the .kps (package source) file to satisfy our build system. You'll want to pull the latest changes (or at least download the revised .kps file from https://github.com/keymanapp/keyboards/blob/master/release/d/dazaga_gourane_karra/source/dazaga_gourane_karra.kps).
It usually takes a few hours before the keyboard is deployed to keyman.com.
